### ISSUE TITLE
Methods in subclassing are biased towards the parent class

### ISSUE THREAD

**Issue by dalyIsaac** [microsoft/pyright#418]
**Describe the bug**
A method residing in a parent class which returns a new instance of class of the calling variable results in the produced variable having the class of the parent.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
``` Python
class A():
    def duplicate(self):
        return type(self)()

class B(A):
    pass

b = B() # type is `B`
bb = b.duplicate() # type is reported to be `A` when it should be `B`
```

**VS Code extension or command-line**

Pyright v1.1.9 running in VS Code

**Comment by erictraut** [comment]
If the `self` input parameter is unannotated, pyright assumes that it is an instance of the class in which the method is defined (in your example, an instance of `A`).

If you want to support the scenario you describe above, you can give it an explicit type annotation using a type variable.

```
_T = TypeVar('_T')

class A():
    def duplicate(self: _T) -> _T:
        return type(self)()
```

This pattern is used in several places within the stdlib type stub files, so I presume that it will work with all python type checkers including mypy.

**Comment by dalyIsaac** [comment]
Thank you ☺️
